### PR TITLE
icinga2_feature: fix module in check_mode

### LINKED ISSUE #39830: Module `icinga2_feature` broken in `check_mode`
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
icinga2_feature

##### ANSIBLE VERSION
```
ansible 2.4.4.0
  config file = /home/ansible/ansible/ansible.cfg
  configured module search path = [u'/home/ansible/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.13 (default, Nov 24 2017, 17:33:09) [GCC 6.3.0 20170516]
```

##### CONFIGURATION
ALLOW_WORLD_READABLE_TMPFILES(/home/ansible/ansible/ansible.cfg) = True
ANSIBLE_PIPELINING(/home/ansible/ansible/ansible.cfg) = True
CACHE_PLUGIN(/home/ansible/ansible/ansible.cfg) = jsonfile
CACHE_PLUGIN_CONNECTION(/home/ansible/ansible/ansible.cfg) = ~/.cache/
CACHE_PLUGIN_TIMEOUT(/home/ansible/ansible/ansible.cfg) = 86400
DEFAULT_BECOME(/home/ansible/ansible/ansible.cfg) = True
DEFAULT_FORKS(/home/ansible/ansible/ansible.cfg) = 10
DEFAULT_GATHERING(/home/ansible/ansible/ansible.cfg) = smart
DEFAULT_HOST_LIST(/home/ansible/ansible/ansible.cfg) = [u'/home/ansible/ansible/hosts']
DEFAULT_MANAGED_STR(/home/ansible/ansible/ansible.cfg) = Ansible managed file, do not edit directly
DEFAULT_REMOTE_USER(/home/ansible/ansible/ansible.cfg) = ansible
DEFAULT_ROLES_PATH(/home/ansible/ansible/ansible.cfg) = [u'/home/ansible/ansible/roles']
DEFAULT_VAULT_PASSWORD_FILE(/home/ansible/ansible/ansible.cfg) = /home/ansible/ansible/open_the_vault.sh
RETRY_FILES_ENABLED(/home/ansible/ansible/ansible.cfg) = False

##### OS / ENVIRONMENT
Debian Stretch GNU Linux

##### SUMMARY
The Ansible module `icinga2_feature` is broken when run in `check_mode`.

##### STEPS TO REPRODUCE

```yaml
- name: Enable Icinga2 features
  icinga2_feature:
    name: "{{ item }}"
    state: present
  with_items:
    - api
    - command
    - ido-mysql
  notify: restart icinga2
```

##### EXPECTED RESULTS
The module should succeed in `check_mode`.

##### ACTUAL RESULTS

```
TASK [icinga2-server : Enable Icinga2 features] *******************************************
task path: /home/ansible/ansible/roles/icinga2-server/tasks/icinga2.yml:32
Using module file /usr/local/lib/python2.7/dist-packages/ansible/modules/monitoring/icinga2_feature.py
<monitor.example.org> ESTABLISH SSH CONNECTION FOR USER: ansible
<monitor.example.org> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/home/ansible/.ansible/cp/6547af9212 monitor.example.org '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-bedifijcwlsgibsuolciioaayzpjnvhp; /usr/bin/python'"'"'"'"'"'"'"'"' && sleep 0'"'"''
<monitor.example.org> (1, '', 'Traceback (most recent call last):\n  File "/tmp/ansible_i6MMOX/ansible_module_icinga2_feature.py", line 127, in <module>\n    main()\n  File "/tmp/ansible_i6MMOX/ansible_module_icinga2_feature.py", line 124, in main\n    Icinga2FeatureHelper(module).manage()\n  File "/tmp/ansible_i6MMOX/ansible_module_icinga2_feature.py", line 76, in manage\n    if re.search("Disabled features:.* %s[ \\n]" % self.module.params["name"], out) \\\n  File "/usr/lib/python2.7/re.py", line 146, in search\n    return _compile(pattern, flags).search(string)\nTypeError: expected string or buffer\n')
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_i6MMOX/ansible_module_icinga2_feature.py", line 127, in <module>
    main()
  File "/tmp/ansible_i6MMOX/ansible_module_icinga2_feature.py", line 124, in main
    Icinga2FeatureHelper(module).manage()
  File "/tmp/ansible_i6MMOX/ansible_module_icinga2_feature.py", line 76, in manage
    if re.search("Disabled features:.* %s[ \n]" % self.module.params["name"], out) \
  File "/usr/lib/python2.7/re.py", line 146, in search
    return _compile(pattern, flags).search(string)
TypeError: expected string or buffer

failed: [monitor.example.org] (item=api) => {
    "changed": false, 
    "failed": true, 
    "item": "api", 
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_i6MMOX/ansible_module_icinga2_feature.py\", line 127, in <module>\n    main()\n  File \"/tmp/ansible_i6MMOX/ansible_module_icinga2_feature.py\", line 124, in main\n    Icinga2FeatureHelper(module).manage()\n  File \"/tmp/ansible_i6MMOX/ansible_module_icinga2_feature.py\", line 76, in manage\n    if re.search(\"Disabled features:.* %s[ \\n]\" % self.module.params[\"name\"], out) \\\n  File \"/usr/lib/python2.7/re.py\", line 146, in search\n    return _compile(pattern, flags).search(string)\nTypeError: expected string or buffer\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE", 
    "rc": 1
}
Using module file /usr/local/lib/python2.7/dist-packages/ansible/modules/monitoring/icinga2_feature.py
<monitor.example.org> ESTABLISH SSH CONNECTION FOR USER: ansible
<monitor.example.org> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/home/ansible/.ansible/cp/6547af9212 monitor.example.org '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-hdfcawsogmmfrhwhmcfyzjciphyotqqw; /usr/bin/python'"'"'"'"'"'"'"'"' && sleep 0'"'"''
<monitor.example.org> (1, '', 'Traceback (most recent call last):\n  File "/tmp/ansible_yR2VVe/ansible_module_icinga2_feature.py", line 127, in <module>\n    main()\n  File "/tmp/ansible_yR2VVe/ansible_module_icinga2_feature.py", line 124, in main\n    Icinga2FeatureHelper(module).manage()\n  File "/tmp/ansible_yR2VVe/ansible_module_icinga2_feature.py", line 76, in manage\n    if re.search("Disabled features:.* %s[ \\n]" % self.module.params["name"], out) \\\n  File "/usr/lib/python2.7/re.py", line 146, in search\n    return _compile(pattern, flags).search(string)\nTypeError: expected string or buffer\n')
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_yR2VVe/ansible_module_icinga2_feature.py", line 127, in <module>
    main()
  File "/tmp/ansible_yR2VVe/ansible_module_icinga2_feature.py", line 124, in main
    Icinga2FeatureHelper(module).manage()
  File "/tmp/ansible_yR2VVe/ansible_module_icinga2_feature.py", line 76, in manage
    if re.search("Disabled features:.* %s[ \n]" % self.module.params["name"], out) \
  File "/usr/lib/python2.7/re.py", line 146, in search
    return _compile(pattern, flags).search(string)
TypeError: expected string or buffer

failed: [monitor.example.org] (item=command) => {
    "changed": false, 
    "failed": true, 
    "item": "command", 
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_yR2VVe/ansible_module_icinga2_feature.py\", line 127, in <module>\n    main()\n  File \"/tmp/ansible_yR2VVe/ansible_module_icinga2_feature.py\", line 124, in main\n    Icinga2FeatureHelper(module).manage()\n  File \"/tmp/ansible_yR2VVe/ansible_module_icinga2_feature.py\", line 76, in manage\n    if re.search(\"Disabled features:.* %s[ \\n]\" % self.module.params[\"name\"], out) \\\n  File \"/usr/lib/python2.7/re.py\", line 146, in search\n    return _compile(pattern, flags).search(string)\nTypeError: expected string or buffer\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE", 
    "rc": 1
}
Using module file /usr/local/lib/python2.7/dist-packages/ansible/modules/monitoring/icinga2_feature.py
<monitor.example.org> ESTABLISH SSH CONNECTION FOR USER: ansible
<monitor.example.org> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/home/ansible/.ansible/cp/6547af9212 monitor.example.org '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-bkbraoemgssmthimioxrkyoqcjxfgrhm; /usr/bin/python'"'"'"'"'"'"'"'"' && sleep 0'"'"''
<monitor.example.org> (1, '', 'Traceback (most recent call last):\n  File "/tmp/ansible__gk_ri/ansible_module_icinga2_feature.py", line 127, in <module>\n    main()\n  File "/tmp/ansible__gk_ri/ansible_module_icinga2_feature.py", line 124, in main\n    Icinga2FeatureHelper(module).manage()\n  File "/tmp/ansible__gk_ri/ansible_module_icinga2_feature.py", line 76, in manage\n    if re.search("Disabled features:.* %s[ \\n]" % self.module.params["name"], out) \\\n  File "/usr/lib/python2.7/re.py", line 146, in search\n    return _compile(pattern, flags).search(string)\nTypeError: expected string or buffer\n')
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible__gk_ri/ansible_module_icinga2_feature.py", line 127, in <module>
    main()
  File "/tmp/ansible__gk_ri/ansible_module_icinga2_feature.py", line 124, in main
    Icinga2FeatureHelper(module).manage()
  File "/tmp/ansible__gk_ri/ansible_module_icinga2_feature.py", line 76, in manage
    if re.search("Disabled features:.* %s[ \n]" % self.module.params["name"], out) \
  File "/usr/lib/python2.7/re.py", line 146, in search
    return _compile(pattern, flags).search(string)
TypeError: expected string or buffer

failed: [monitor.example.org] (item=ido-mysql) => {
    "changed": false, 
    "failed": true, 
    "item": "ido-mysql", 
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible__gk_ri/ansible_module_icinga2_feature.py\", line 127, in <module>\n    main()\n  File \"/tmp/ansible__gk_ri/ansible_module_icinga2_feature.py\", line 124, in main\n    Icinga2FeatureHelper(module).manage()\n  File \"/tmp/ansible__gk_ri/ansible_module_icinga2_feature.py\", line 76, in manage\n    if re.search(\"Disabled features:.* %s[ \\n]\" % self.module.params[\"name\"], out) \\\n  File \"/usr/lib/python2.7/re.py\", line 146, in search\n    return _compile(pattern, flags).search(string)\nTypeError: expected string or buffer\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE", 
    "rc": 1
}

```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Fixes: #39830

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/monitoring/icinga2_feature.py
test/units/modules/monitoring/test_icinga2_feature.py

##### ANSIBLE VERSION

```
2.6devel
```

### CODE REVIEW

File: lib/ansible/modules/monitoring/icinga2_feature.py [hunk 1]
```diff
@@ -21,49 +25,57 @@
...
 
 class Icinga2FeatureHelper:
     def __init__(self, module):
         self.module = module
         self._icinga2 = module.get_bin_path('icinga2', True)
+        self.feature_name = self.module.params['name']
+        self.state = self.module.params['state']
 
     def _exec(self, args):
         if not self.module.check_mode:
```
Comment (mejo-) [inline lib/ansible/modules/monitoring/icinga2_feature.py]: This condition is the problem. It leads to `icinga2 feature list` never beeing executed in `check_mode`, which results in empty `out`. The condition should be removed alltogether. `self._exec` is only invoked twice anyway: first to run `icinga2 feature list` (should be done regardless of `check_mode` or not) and second at a place in code that is never reached in `check_mode` anyway.

**Comment by Akasurde** [comment]
@mejo- I updated as per reviews, could you please check again ?

**Comment by mejo-** [comment]
Looks great now and works as expected.

**Review by mejo- — APPROVED** [review]
I tested the new version and it works as expected. Looks good to me.

**Comment by Akasurde** [comment]
@mejo- Thanks for the reviews and testing the module.

**Pull request merged**